Add instructions to securely manage token (#2322)
Qiskit/documentation: docs/guides/setup-channel.mdx

      The Instances section in your [IBM Quantum account page](https://quantum.ibm.com/account) lists the instances that you can access.
    </Admonition>

> 1. <span id="save-account"></span>Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account). **If you are on a trusted machine,** use the `save_account()` method to save your credentials locally. (Skip to the next step if you are not using a trusted machine to authenticate to IBM Quantum Platform.) To use `save_account()`, first activate your Python virtual environment (see the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up) and run Python to open a read-eval-print loop (REPL), then enter the following:

Reviewer: I think it's useful to call it a REPL so people can search online for a tutorial if they're not familiar. I would word it like this:

> With your virtual environment activated, run `python` in your shell to open a REPL (read-eval-print loop). Then, enter the following:

Note that most devs are more familiar with the term REPL than `read-eval-print loop`, so I inverted the parantheses

> 1. <span id="save-account"></span>Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account). **If you are on a trusted machine,** use the `save_account()` method to save your credentials locally. (Skip to the next step if you are not using a trusted machine to authenticate to IBM Quantum Platform.) To use `save_account()`, first activate your Python virtual environment (see the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up) and run Python to open a read-eval-print loop (REPL), then enter the following:

Author: ```suggestion
1. <span id="save-account"></span>Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account). **If you are on a trusted machine,** use the `save_account()` method to save your credentials locally. (Skip to the next step if you are not using a trusted machine to authenticate to IBM Quantum Platform.) To use `save_account()`, first activate your Python virtual environment (see the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up) and run Python in your shell to open a REPL (read-eval-print loop). Then, enter the following:
```

      The Instances section in your [IBM Quantum account page](https://quantum.ibm.com/account) lists the instances that you can access.
    </Admonition>

1. Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account), and activate your Python virtual environment.  See the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up. 

>     <span id="save-account"></span>**If you are working in a trusted Python environment (such as on a personal laptop or workstation),** use the `save_account()` method to save your credentials locally. ([Skip to the next step](#untrusted-machine) if you are not using a trusted environment, such as a shared or public computer, to authenticate to IBM Quantum Platform.) To use `save_account()`, run Python in your virtual environment to open a read-eval-print loop (REPL), then enter the following:

Reviewer 2: Can we also link to the [save_account](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.QiskitRuntimeService#save_account) API in the first mention?

>     <span id="save-account"></span>**If you are working in a trusted Python environment (such as on a personal laptop or workstation),** use the `save_account()` method to save your credentials locally. ([Skip to the next step](#untrusted-machine) if you are not using a trusted environment, such as a shared or public computer, to authenticate to IBM Quantum Platform.) To use `save_account()`, run Python in your virtual environment to open a read-eval-print loop (REPL), then enter the following:

Author: done in 2bb59bfa7c887b044ca055a058d31c9849992e9c

    ```python
    from qiskit_ibm_runtime import QiskitRuntimeService

    QiskitRuntimeService.save_account(
      token="<your-token">,
      channel="ibm_quantum", # `channel` distinguishes between different account types
      overwrite=True # Use `overwrite=True` if you're updating your token

>     )

Reviewer 2: If we are recommending people to use REPL, then the code snippet should be structured in a way that can be copy&pasted into the REPL and the token modified easily. So I'd recommend changing this to 3 blocks of code:

block 1:
```python 
from qiskit_ibm_runtime import QiskitRuntimeService
```

block 2:
```python
token = "<your-token>"
```

block 3:
```python
QiskitRuntimeService.save_account(token=token, channel="ibm_quantum")
```

Personally I don't like `overwrite=True`. The current code block has `token="<your-token">`, which presumably meant to be `token="<your-token>",`. If this was fixed, and I copy&paste the block into my REPL, it'd overwrite my saved account with `<your-token>` as the token!

Author: removed the overwrite and split into two blocks f20da5afcad797f65a9b408462abdde613390f0a and 84de35b3570afb1dee53952ed6c74136a7603a79

>     )

Author: done in f20da5afcad797f65a9b408462abdde613390f0a

    ```

    Close out of the REPL with `exit()`. From now on, whenever you need to authenticate to the service, you can load your credentials with `QiskitRuntimeService()`.

    ```python
    # Load saved credentials
    service = QiskitRuntimeService()
    ```

      * If you have saved credentials for both an IBM Quantum Platform account and an IBM Cloud account, IBM Cloud is the default account used when you initialize your account, unless you specify `set_as_default=True` in your IBM Quantum Platform account when you use the `save_account()` method.
      * If you are saving multiple accounts per channel, consider using the `name` parameter to differentiate them.
      * Credentials are saved to `$HOME/.qiskit/qiskit-ibm.json`.  Do not manually edit this file.

    <Admonition type="warning">

>       **Protect your API token!** Never include your token in plain text in a notebook file. If you are working in a trusted Python environment (such as on a personal laptop or workstation), use the `save_account()` method described in the previous step to store your credentials, keeping in mind that your token will still be stored as plain text on your local drive. If you are executing code in a Python environment that is not trusted, see the [instructions in the next step](#untrusted-machine).

Reviewer 2: This warning should be higher up - e.g. before any code snippets. Otherwise people would likely copy&paste the code into their notebook before getting to this warning.

>       **Protect your API token!** Never include your token in plain text in a notebook file. If you are working in a trusted Python environment (such as on a personal laptop or workstation), use the `save_account()` method described in the previous step to store your credentials, keeping in mind that your token will still be stored as plain text on your local drive. If you are executing code in a Python environment that is not trusted, see the [instructions in the next step](#untrusted-machine).

Author: done in 0e18cd37b89489d3fcfb0d38a41d5163e9134178

    </Admonition>

> 1. <span id="untrusted-machine"></span>**Avoid executing code on an external or random cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by rotating it on the [IBM Quantum Platform dashboard](https://quantum.ibm.com/) (click the refresh button in the API token field). To initialize the service in this situation, you can use code like the following:

Reviewer 3: ```suggestion
1. <span id="untrusted-machine"></span>**Avoid executing code on an external cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by rotating it on the [IBM Quantum Platform dashboard](https://quantum.ibm.com/) (click the refresh button in the API token field). To initialize the service in this situation, you can use code like the following:
```

    ```python
    from qiskit_ibm_runtime import QiskitRuntimeService
  
    # Refresh your API token on the dashboard after using the code as follows:
    service = QiskitRuntimeService(channel="ibm_quantum", token="<MY_IBM_QUANTUM_TOKEN>")
    ```

    <Admonition type="warning">

>       When sharing code with others, ensure that your API token is not embedded directly within the Python script in plain text. Instead, share the script without the token and provide instructions for securely setting it up.

Reviewer 3: ```suggestion
      When sharing code with others, ensure that your API token is not embedded directly within the Python script. Instead, share the script without the token and provide instructions for securely setting it up.
```

Shouldn't include the token at all.

Reviewer 2: I'd also add `.. Python script or notebook`... since notebooks are the most common way for people to share.

Author: resolved by 1f3d40de4d4fdcde7f47199b13ace582c4466670 and 0e18cd37b89489d3fcfb0d38a41d5163e9134178

1. Retrieve your IBM Quantum token from the [IBM Quantum account page](https://quantum.ibm.com/account).

1. <span id="save-account"></span>Authenticate to the service by calling `QiskitRuntimeService` with your IBM Quantum API token and CRN. [Set up an environment variable](#securely-manage) rather than copying and pasting your API token into your code.

   ```python
   from qiskit_ibm_runtime import QiskitRuntimeService

>    service = QiskitRuntimeService(channel="ibm_quantum", token=<your-environment-variable>)

Reviewer 4: If we're telling them to set it as IQP_API_TOKEN, why not just use that instead of <your-environment-variable>?

>    service = QiskitRuntimeService(channel="ibm_quantum", token=<your-environment-variable>)

Reviewer 4: I tried this but got an error"...IBMNotAuthorizedError: '401 Client Error: Unauthorized for url: https://auth.quantum-computing.ibm.com/api/users/loginWithToken. Login failed., Error code: 3446.'" 

Maybe I set something wrong?  I found instructions to update .bashrc and added my environment variable, then restarted bash and used it to start jupyter lab.  Then I put this code in a notebook: 

```
from qiskit_ibm_runtime import QiskitRuntimeService
 
service = QiskitRuntimeService(channel="ibm_quantum", token=("IQP_API_TOKEN"))
```

>    service = QiskitRuntimeService(channel="ibm_quantum", token=<your-environment-variable>)

Reviewer: right now you are setting the token as the hardcoded string "IQP_API_TOKEN". It would need to be `os.environ.pop("IQP_API_TOKEN")`

>    service = QiskitRuntimeService(channel="ibm_quantum", token=<your-environment-variable>)

Reviewer 2: If you have already done `save_account()`, then `QiskitRuntimeService()` is sufficient. 

For people who don't want to do `save_account()`, presumably they wouldn't want to save it in .bashrc either, since both store the credentials in plain text locally. In this case they'd have to manually code `QiskitRuntimeService(channel=..., token=..., instance=...)`. 

I wouldn't recommend doing `QiskitRuntimeService(token=os.environ.pop("IQP_API_TOKEN"))` because `save_account` saves more than just the token. And the code doesn't support mixing the profile info with parameter info. For example: 

```python
# This saves to the profile, including the instance
QiskitRuntimeService.save_account(channel=..., token=..., instance="h1/g1/p1") 

# This loads from the profile, including the instance
service = QiskitRuntimeService()
backend = service.backend("ibm_fez")
Sampler(backend).run(...)  #  <-- this will run using the instance h1/g1/p1

# This does NOT load from the profile and hence not the instance
service = QiskitRuntimeService(token=os.environ.pop("IQP_API_TOKEN"))
backend = service.backend("ibm_fez")
Sampler(backend).run(...)  #  <-- this will use a random instance
```

   ```

>    Or, optionally use the `save_account()` method to save your credentials for easy access later on, before initializing the service. *Note that account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.*

Reviewer 4: ```suggestion
   Or, optionally use the `save_account()` method to save your credentials for easy access later on, before initializing the service. *Note that account credentials are saved in plain text. To more securely manage your token, it is recommended that you use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.*
```

    * The `channel` parameter allows you to distinguish between different account types. When initializing the account, IBM Cloud is the default account used if have saved credentials for an IBM Quantum Platform and an IBM Cloud account.

    <Admonition type="caution">

>       Account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.

Reviewer 4: ```suggestion
      Account credentials are saved in plain text. To more securely manage your token, it is recommended that you use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.
```

    * The `channel` parameter allows you to distinguish between different account types. When initializing the account, IBM Cloud is the default account used if have saved credentials for an IBM Quantum Platform and an IBM Cloud account.

    <Admonition type="caution">
      Account credentials are saved in plain text, so only do so if you are using a trusted device.
      Account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.
    </Admonition>

>     <Admonition type="note">

Reviewer: Why remove this?

    IBM Cloud is the default account used if you don't specify a different channel or account name.
    </Admonition>

>     <CodeAssistantAdmonition

Reviewer: This is removed because it's not secure, right? Indeed, you should not send a credential to Qiskit Code Assistant

Author: Removing this was an accident! I will restore the one above but delete the code assistant one -  @Eric-Arellano verifying I've got it right?

Reviewer: Yeah, that sounds good. Thanks!

    <Admonition type="note">
      The Instances section in your [IBM Quantum account page](https://quantum.ibm.com/account) lists the instances that you can access.
    </Admonition>

1. Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account).

1. Because of the security risks posed by executing code that contains your API token in plain text, the recommended authentication method is to first create an environment variable for your API token, but only do so **if you are working in a trusted Python environment (such as on a personal laptop or workstation).** [Skip to the next step](#rest-untrusted) if you are not using a trusted environment, such as a shared or public computer.

>     The following demonstrates creating an environment variable:

Reviewer 2: for a trusted device, we can use the original wording - _Set the IQP_API_TOKEN environment variable in your system. You can do this by adding the following line to your shell profile (e.g., .bashrc, .zshrc) or by setting it directly in your terminal:_

Author: resolved by 1f3d40de4d4fdcde7f47199b13ace582c4466670

    import os
    import requests
    
    reqUrl = "https://api.quantum-computing.ibm.com/runtime/jobs?limit=10&offset=0&exclude_params=true"
    
    headersList = {
      "Accept": "application/json",
      "Authorization": "Bearer os.environ.pop(IQP_API_TOKEN)"
    }
    
    payload = ""
    
    response = requests.request("GET", reqUrl, data=payload,  headers=headersList)
    
    print(response.json())
    ```

    Note that when creating an environment variable, your API token is still stored locally in plain text, and should be safeguarded.

1. <span id="rest-untrusted"></span>**Avoid executing code on an external or random cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by rotating it on the [IBM Quantum Platform dashboard](https://quantum.ibm.com/) (click the refresh button in the API token field). To initialize the service in this situation, use your API token directly: 

    ```python
    # Refresh your API token on the dashboard after using the code as follows:
    "Authorization": "Bearer <your-token>" 
    ```

    <Admonition type="warning">

>       When sharing code with others, ensure that your API token is not embedded directly within the Python script in plain text. Instead, share the script without the token and provide instructions for securely setting it up.

Reviewer 3: ```suggestion
      When sharing code with others, ensure that your API token is not embedded directly within the Python script. Instead, share the script without the token and provide instructions for securely setting it up.
```

    </Admonition>

1. Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account).

1. Because of the security risks posed by executing code that contains your API token, the recommended authentication method is to first create an environment variable for your API token, but only do so **if you are working in a trusted Python environment (such as on a personal laptop or workstation).** [Skip to the next step](#rest-untrusted) if you are not using a trusted environment, such as a shared or public computer.

    To set the IQP_API_TOKEN environment variable in your system, you can add the following line to your shell profile (for example, .bashrc or .zshrc) or by setting it directly in your terminal:

    ```shell
    export IQP_API_TOKEN=<your-token>
    ```
    When you invoke the environment variable in your code, include `import os`, as in this example request:

    ```python
    import os
    import requests
    
    reqUrl = "https://api.quantum-computing.ibm.com/runtime/jobs?limit=10&offset=0&exclude_params=true"
    

>     headersList = {
>       "Accept": "application/json",
>       "Authorization": "Bearer os.environ.pop(IQP_API_TOKEN)"
>     }

Reviewer 2: The code to get the token from env variable needs to actually run, e.g. 

```suggestion
    headersList = {
      "Accept": "application/json",
      "Authorization": f"Bearer {os.environ['IQP_API_TOKEN']}"
    }
```

Also don't use `pop()` since it removes the token from the env variable.

    <Admonition type="note">
      The Instances section in your [IBM Quantum account page](https://quantum.ibm.com/account) lists the instances that you can access.
    </Admonition>

1. Retrieve your API token from the [IBM Quantum account page](https://quantum.ibm.com/account).

1. Because of the security risks posed by executing code that contains your API token, the recommended authentication method is to first create an environment variable for your API token, but only do so **if you are working in a trusted Python environment (such as on a personal laptop or workstation).** [Skip to the next step](#rest-untrusted) if you are not using a trusted environment, such as a shared or public computer.

    To set the IQP_API_TOKEN environment variable in your system, you can add the following line to your shell profile (for example, .bashrc or .zshrc) or by setting it directly in your terminal:

    ```shell
    export IQP_API_TOKEN=<your-token>
    ```
    When you invoke the environment variable in your code, include `import os`, as in this example request:

    ```python
    import os
    import requests
    

>     reqUrl = "https://api.quantum-computing.ibm.com/runtime/jobs?limit=10&offset=0&exclude_params=true"
>     
>     headersList = {
>       "Accept": "application/json",
>       "Authorization": f"Bearer {os.environ['IQP_API_TOKEN']}"
>     }
>     
>     payload = ""
>     
>     response = requests.request("GET", reqUrl, data=payload,  headers=headersList)

Reviewer: Nit to not annoy Pythonistas lol:

```suggestion
    req_url = "https://api.quantum-computing.ibm.com/runtime/jobs?limit=10&offset=0&exclude_params=true"

    headers_list = {
      "Accept": "application/json",
      "Authorization": f"Bearer {os.environ['IQP_API_TOKEN']}"
    }
    
    payload = ""
    
    response = requests.request("GET", req_url, data=payload,  headers=headers_list)
```

1. <span id="rest-untrusted"></span>**Avoid executing code on an untrusted machine or an external cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by rotating it on the [IBM Quantum Platform dashboard](https://quantum.ibm.com/) (click the refresh button in the API token field). To initialize the service in this situation, use your API token directly: 

    ```python
    # Refresh your API token on the dashboard after using the code as follows:
    "Authorization": "Bearer <your-token>" 
    ```

    <Admonition type="caution">
      When sharing code with others, ensure that your API token is not embedded directly within the Python script. Instead, share the script without the token and provide instructions for securely setting it up.

      If you accidentally share your token with someone or include it in version control like Git, immediately revoke your token by expiring it on the [IBM Quantum Platform dashboard](https://quantum.ibm.com/) (click the refresh button in the API token field) to reduce risk.
    </Admonition>

#### Use a temporary access token

> 1. Optionally obtain a temporary access token by supplying your API token. This is especially useful if you would like control over tokens, such as token invalidation.

Reviewer: ```suggestion
1. Optionally obtain a temporary access token by supplying your API token. This is especially useful if you would like control over tokens, such as token invalidation. Follow the same precautions as above with securely using the environment variable.
```

(I think we aren't supposed to say "above"?)

--

You also will want to update the code to use `import os` and read in the environment variable like you did above. 

While you're fixing this code, can you please rename `input` to be `papyload`, and have a space between the equals like `payload = {`? `input` is a builtin value in Python and we're overwriting it, which is a bad practice and messes up code highlighting

Author: done in 065dd390187b427bc9aa7736439afcf90691622e

    1. Find your API key. From the [API keys page](https://cloud.ibm.com/iam/apikeys), view or create your API key, then copy it to a secure location so you can use it for authentication.
    2. Find your Cloud Resource Name (CRN). Open the [Instances page](https://cloud.ibm.com/quantum/instances) and click your instance. In the page that opens, click the icon to copy your CRN. Save it in a secure location so you can use it for authentication.

1. <span id="cloud-save"></span>Authenticate to the service by calling `QiskitRuntimeService` with your saved credentials or with your IBM Cloud API key and CRN:

   ```python
   from qiskit_ibm_runtime import QiskitRuntimeService

   service = QiskitRuntimeService(channel="ibm_cloud", token="<IBM Cloud API key>", instance="<IBM Cloud CRN>")
   ```

>    You can optionally use the `save_account()` method to save your credentials for easy access later on, before initializing the service. *Note that account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.* 

Reviewer 4: ```suggestion
   You can optionally use the `save_account()` method to save your credentials for easy access later on, before initializing the service. *Note that account credentials are saved in plain text. To more securely manage your token, it is recommended that you use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.* 
```

    * Credentials are saved to `$HOME/.qiskit/qiskit-ibm.json`.  Do not manually edit this file.
    * If you don't save your credentials, you must specify them every time you start a new session.

    <Admonition type="caution">
      Account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.
    </Admonition>

6. Test your setup.  Ensure that you can connect to the service:

```python
  from qiskit_ibm_runtime import QiskitRuntimeService

  service = QiskitRuntimeService()
```

<span id="securely-manage"></span>
## Securely manage your token

> To store sensitive information like account credentials more securely, it’s a good practice to use environment variables instead of saving them in plain text. Here’s how to modify your code to load the API token from an environment variable.

Reviewer: Should this be something like "The following describes how"?

Author: got it ca16e6962f24e8c93d6e703dcf3b02ffefec076b

1. Set the environment variable.

Set the `IQP_API_TOKEN` environment variable in your system. You can do this by adding the following line to your shell profile (for example, `.bashrc`, `.zshrc`) or by setting it directly in your terminal:

```bash
export IQP_API_TOKEN="<YOUR_IQP_API_TOKEN>"
```

2. Modify the code to retrieve the token from the environment variable.

Update your code to retrieve the token from the environment variable instead of saving it directly:

```python
import os

> from qiskit_ibm_catalog import QiskitFunctionsCatalog

Reviewer: This example is only relevant for Functions, which not everyone uses. Should this code sample be for Runtime? And then Functions is a subsection?

Author: @pandasa123 we could use a Runtime code sample for the setup-channel page, and then use the one you provided for the Functions page (adding a separate section there). Sound ok?

Reviewer: But on https://docs.quantum.ibm.com/guides/install-qiskit we don't say to install qiskit-ibm-catalog, and there is no reason to install it unless you're using Functions.

Author: @pandasa123 ⬆️

Reviewer 4: > Actually, just spoke to Sanket in-person,
Just casually making us jealous...

    </Admonition>

6. Test your setup.  Ensure that you can connect to the service:

```python
  from qiskit_ibm_runtime import QiskitRuntimeService

  service = QiskitRuntimeService()
```

<span id="securely-manage"></span>
## Securely manage your token

To store sensitive information like account credentials more securely, it’s a good practice to use environment variables instead of saving them in plain text. The following demonstrates how to modify your code to load the API token from an environment variable.

1. Set the environment variable.

Set the `IQP_API_TOKEN` environment variable in your system. You can do this by adding the following line to your shell profile (for example, `.bashrc`, `.zshrc`) or by setting it directly in your terminal:

```bash
export IQP_API_TOKEN="<YOUR_IQP_API_TOKEN>"
```

2. Modify the code to retrieve the token from the environment variable.

Update your code to retrieve the token from the environment variable instead of saving it directly:

```python
import os
from qiskit_ibm_catalog import QiskitFunctionsCatalog

> # Gather the token from your environment
> iqp_api_token = os.getenv("IQP_API_TOKEN")

Reviewer: ```suggestion
# Gather the token from your environment variables.
# We use `.pop()` to ensure any subprocesses started by
# our program cannot access the value.
iqp_api_token = os.environ.pop("IQP_API_TOKEN")
```

    * Credentials are saved to `$HOME/.qiskit/qiskit-ibm.json`.  Do not manually edit this file.
    * If you don't save your credentials, you must specify them every time you start a new session.

    <Admonition type="caution">

>       Account credentials are saved in plain text. To more securely manage your token, it is recommended to use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.

Reviewer 4: ```suggestion
      Account credentials are saved in plain text. To more securely manage your token, it is recommended that you use an environment variable; see the [Securely manage your token](#securely-manage) section for more information.
```

    2. Find your Cloud Resource Name (CRN). Open the [Instances page](https://cloud.ibm.com/quantum/instances) and click your instance. In the page that opens, click the icon to copy your CRN. Save it in a secure location so you can use it for authentication.

1. <span id="cloud-save"></span>**If you are working in a trusted Python environment (such as on a personal laptop or workstation),** use the `save_account()` method to save your credentials locally. ([Skip to the next step](#cloud-untrusted) if you are not using a trusted environment, such as a shared or public computer, to authenticate to IBM Cloud.) To use `save_account()`, activate a Python virtual environment (see the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up). Then, run Python in your virtual environment to open a read-eval-print loop (REPL), and enter the following:

    ```python
    from qiskit_ibm_runtime import QiskitRuntimeService

    service = QiskitRuntimeService(
    channel="ibm_cloud",

>     token="<IBM Cloud API key>", # Your token is confidential.

Reviewer 2: Same as my comment for IQP setup, I'd separate out 

```python
token = <IBM Cloud API key>
```

from the rest, so it's easy to edit in REPL.

Author: done in 6f5da2e749e1e5648bdf8fa9cde467cd7c756be5

      * Credentials are saved to `$HOME/.qiskit/qiskit-ibm.json`.  Do not manually edit this file.
      * If you don't save your credentials, you must specify them every time you start a new session.

1. <span id="cloud-untrusted"></span>**Avoid executing code on an external cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by deleting it on the [API keys page](https://cloud.ibm.com/iam/apikeys) and creating a new one. To initialize the service in this situation, you can use code like the following:

    ```python
    from qiskit_ibm_runtime import QiskitRuntimeService
    
    service = QiskitRuntimeService(
      channel="ibm_cloud", 
      # Delete your token on the API keys page after entering this code:
      token="<IBM Cloud API key>", 
      instance="<IBM Cloud CRN>"
      )
    ```
    <Admonition type="caution">
      When sharing code with others, ensure that your API token is not embedded directly within the Python script. Instead, share the script without the token and provide instructions for securely setting it up.

>     </Admonition>

Reviewer 2: Same comment as before - that this should be higher up. I assume you'll just copy&paste. I like the bolded `Protect your API key!` in the other warning.

Author: done in 0be84fc58abae1f333ecfaac1043e739a6e659ce

    2. Find your Cloud Resource Name (CRN). Open the [Instances page](https://cloud.ibm.com/quantum/instances) and click your instance. In the page that opens, click the icon to copy your CRN. Save it in a secure location so you can use it for authentication.

1. <span id="cloud-save"></span>**If you are working in a trusted Python environment (such as on a personal laptop or workstation),** use the `save_account()` method to save your credentials locally. ([Skip to the next step](#cloud-untrusted) if you are not using a trusted environment, such as a shared or public computer, to authenticate to IBM Cloud.) To use `save_account()`, activate a Python virtual environment (see the [installation instructions](/guides/install-qiskit#local) if you do not already have a virtual environment set up). Then, run Python in your virtual environment to open a read-eval-print loop (REPL), and enter the following:

    ```python
    token = "<your-token>"
    ```

    ```python
    service = QiskitRuntimeService(

>     channel="ibm_cloud", # `channel` distinguishes between different account types
>     token=token, # Your token is confidential.
>     # Take care not to share your token in public code.
>     instance="<IBM Cloud CRN>",
>     name="<account-name>",

Reviewer: ```suggestion
        channel="ibm_cloud", # `channel` distinguishes between different account types
        token=token, # Your token is confidential.
        # Take care not to share your token in public code.
        instance="<IBM Cloud CRN>",
        name="<account-name>",
```

      * Credentials are saved to `$HOME/.qiskit/qiskit-ibm.json`.  Do not manually edit this file.
      * If you don't save your credentials, you must specify them every time you start a new session.

1. <span id="cloud-untrusted"></span>**Avoid executing code on an external cloud Python environment to minimize security risks.** If you must use an untrusted environment (on, for example, a public computer), change your API token after each use by deleting it on the [API keys page](https://cloud.ibm.com/iam/apikeys) and creating a new one. 

    <Admonition type="caution">

>       **Protect your API token!** Never include your token in source code, Python script, or notebook file. When sharing code with others, ensure that your API token is not embedded directly within the Python script. Instead, share the script without the token and provide instructions for securely setting it up.

Reviewer: The Functions guide should be updated to use this wording. Additionally, if you take my recommendation for saying what to do if you accidentally share it, then this file will need that same change.
